test/test_autograd.py

```diff
@@ -9329,6 +9329,47 @@ def train_fn_grad(x):
         # be accumulate to the same place and should be the same
         self._run_py_multithread_fn(train_fn_grad, (x,))
 
+    def test_multithread_saved_tensors_hooks(self):
+        def pack(x):
+            warnings.warn("pack")
+            return x
+
+        def registers_hooks_for_each_thread():
+            try:
+                torch.autograd.graph.set_saved_tensors_default_hooks(pack, lambda x: x)
+                x = torch.ones(5, 5, requires_grad=True)
+                warnings.simplefilter('always')
+                with warnings.catch_warnings(record=True) as w:
+                    y = x * x
+                    # should raise two warnings from x being saved twice
+                    self.assertEqual(len(w), 2)
+            finally:
+                torch.autograd.graph.reset_saved_tensors_default_hooks()
+            y.sum().backward()
+
+    def test_dataparallel_saved_tensors_hooks(self):
+        def pack(x):
+            warnings.warn("pack")
+            return x
+
+        _self = self
+
+        class Model(torch.nn.Module):
+            def forward(self, x):
+                warnings.simplefilter('always')
+                with warnings.catch_warnings(record=True) as w:
+                    y = x * x
+                    _self.assertEqual(len(w), 2)
```

Comment:
This test is currently failing. 
It seems like the default hooks are not propagated through `DataParallel`.

Reply:
TLS is not propagate through DataParallel :(

Reply:
That's actually a good test to make sure that even though you set the hooks in the calling thread, they are not set in the DP worker threads :D

Reply:
Oh I see, I got confused because of the snippet below.
But now I'm realizing that GradMode is handled at the Python level in `torch/nn/parallel/parallel_apply.py` !

```python
import torch

class Model(nn.Module):
    def forward(self, x):
        y = x * x
        print(y.requires_grad)

x = torch.ones(5, 5, requires_grad=True)
model = nn.DataParallel(Model())

model(x)
with torch.no_grad():
    model(x)
```

Output:
```
True
True
False
False
```